designs/multi-cluster.md

```diff
@@ -0,0 +1,376 @@
+}
+client := cl.GetClient()
+```
+
+Due to `request.ClusterAwareRequest`, changes to the controller builder process are minimal:
+
+```golang
+// previous
+builder.TypedControllerManagedBy[reconcile.Request](mgr).
+	Named("single-cluster-controller").
+	For(&corev1.Pod{}).
+	Complete(reconciler)
+
+// new
+builder.TypedControllerManagedBy[reconcile.ClusterAwareRequest](mgr).
+	Named("multi-cluster-controller").
+	For(&corev1.Pod{}).
+	Complete(reconciler)
+```
+
+The builder will chose the correct `EventHandler` implementation for both `For` and `Owns`
+depending on the `request` type used.
+
+With the described changes (use `GetCluster(ctx, req.ClusterName)`, making `reconciler`
+a `TypedFunc[reconcile.ClusterAwareRequest]`) an existing controller will automatically act as
+*uniform multi-cluster controller* if a cluster provider is configured.
+It will reconcile resources from cluster `X` in cluster `X`.
+
+For a manager with `cluster.Provider`, the builder _SHOULD_ create a controller
+that sources events **ONLY** from the provider clusters that got engaged with
+the controller.
+
+Controllers that should be triggered by events on the hub cluster can continue
+to use `For` and `Owns` and explicitly pass the intention to engage only with the
+"default" cluster (this is only necessary if a cluster provider is plugged in):
+
+```golang
+builder.NewControllerManagedBy(mgr).
+    WithOptions(controller.TypedOptions{
+        EngageWithDefaultCluster: ptr.To(true),
+        EngageWithProviderClusters: ptr.To(false),
+    }).
+    For(&appsv1.Deployment{}).
+	Owns(&v1.ReplicaSet{}).
+    Complete(reconciler)
+```
+
+## User Stories
+
+### Controller Author with no interest in multi-cluster wanting to old behaviour.
+
+- Do nothing. Controller-runtime behaviour is unchanged.
+
+### Multi-Cluster Integrator wanting to support cluster managers like Cluster API or kind
+
+- Implement the `cluster.Provider` interface, either via polling of the cluster registry
+  or by watching objects in the hub cluster.
+- For every new cluster create an instance of `cluster.Cluster` and call `mgr.Engage`.
+
+### Multi-Cluster Integrator wanting to support apiservers with logical cluster (like kcp)
+
+- Implement the `cluster.Provider` interface by watching the apiserver for logical cluster objects
+  (`LogicalCluster` CRD in kcp).
+- Return a facade `cluster.Cluster` that scopes all operations (client, cache, indexers) 
+  to the logical cluster, but backed by one physical `cluster.Cluster` resource.
+- Add cross-cluster indexers to the physical `cluster.Cluster` object.
+
+### Controller Author without self-interest in multi-cluster, but open for adoption in multi-cluster setups
+
+- Replace `mgr.GetClient()` and `mgr.GetCache` with `mgr.GetCluster(req.ClusterName).GetClient()` and `mgr.GetCluster(req.ClusterName).GetCache()`.
+- Make manager and controller plumbing vendor'able to allow plugging in multi-cluster provider and BYO request type.
```

Comment:
Is the idea that something like cert-manager would decide on startup which cluster provider is used and can then only work with one cluster provider at a time?

Phrased differently. Do we also have to support multiple cluster providers at the same time?

Reply:
- One provider per manager at a time.
- Low friction to make reconcilers uniform multi-cluster capable (this basically means using the cluster-enabled request and to call `mgr.GetCluster(name)` instead of accessing cluster methods directly).
- If a controller project wants to add support for a number of provider in their repository, this is fine, but not necessarily the goal.
- Instead it should be easy to instantiate the controllers from an alternative main.go with a provider of your choice.